checker/src/block_visitor.rs

```diff
@@ -2208,15 +2223,38 @@ impl<'block, 'analysis, 'compilation, 'tcx, E>
     ) -> Rc<AbstractValue> {
         let result;
         if let rustc_middle::ty::ConstKind::Value(ConstValue::Scalar(Scalar::Raw {
-            data, ..
+            data,
+            size,
+            ..
         })) = &literal.val
         {
+            let enum_def = ty.ty_adt_def().unwrap();
+            let variant_tag = enum_def
+                .variants
+                .indices()
+                .find_map(|idx| {
+                    let variant_discr = enum_def.discriminant_for_variant(self.bv.tcx, idx);
+                    let variant_tag = variant_discr.val; // The variant tag can be explicitly defined in fieldless enums.
+                    let variant_data = match size {
+                        1 => u128::from(variant_tag as u8),
+                        2 => u128::from(variant_tag as u16),
+                        4 => u128::from(variant_tag as u32),
+                        8 => u128::from(variant_tag as u64),
+                        _ => variant_tag,
+                    };
+                    if variant_data == *data {
+                        Some(variant_tag)
+                    } else {
+                        None
+                    }
+                })
+                .unwrap_or(*data); // todo: Deal with optimized layouts for enum types like Option<char>.
```

Comment:
I have a hard time understanding this code. It seems that the point here is to do sign extension for negative numbers. It boggles my mind that it has to be this complicated. Why is variant_discr.val not sufficient in all cases? Some comments with illustrative examples will be really helpful.

Reply:
Added some comments on the complication.

Reply:
Wouldn't tcx.sign_extend be useful to simplify the code?

Reply:
I checked the source code of rustc and found out that we need to `sign_extend` and then `truncate` here.

Instead of just adding the workaround, I figured out that it would better to just implement the logic to deal with enum encoding.